test/tests/selftest/pbs_config_sched.py

```diff
@@ -0,0 +1,133 @@
+# Free Software Foundation, either version 3 of the License, or (at your
+# option) any later version.
+#
+# OpenPBS is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public
+# License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# PBS Pro is commercially licensed software that shares a common core with
+# the OpenPBS software.  For a copy of the commercial license terms and
+# conditions, go to: (http://www.pbspro.com/agreement.html) or contact the
+# Altair Legal Department.
+#
+# Altair's dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of OpenPBS and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair's trademarks, including but not limited to "PBS™",
+# "OpenPBS®", "PBS Professional®", and "PBS Pro™" and Altair's logos is
+# subject to Altair's trademark licensing policies.
+
+from tests.selftest import *
+import re
+
+
+class TestSchedConfig(TestSelf):
+    """
+    Test setting values in sched_config file.
+    """
+
+    def set_and_get(self, confs, validate=False):
+        """
+        Change settings in scheduler config file and fetch the new
+        file contents
+        """
+        sched = self.our_sched
+        sched.set_sched_config(confs, apply=True, validate=validate)
+        cfile = sched.sched_config_file
+        clines = self.du.cat(sched.hostname, cfile, sudo=True,
+                             level=logging.DEBUG2)['out']
+        slines = [x for x in clines if re.match(r'[\w]', x)]
+        return slines
+
+    def get_a_setting(self, slines, key, squash=True):
+        """"
+        Extract lines changing a particular setting from a list of lines
+        """
+        vals = []
+        key_colon = key + ':'
+        for line in slines:
+            if line.startswith(key_colon):
+                value = line[len(key_colon):].strip()
+                if squash:
+                    value = ' '.join(value.split())
+                vals.append(value)
+        return vals
+
+    def test_set_sched_config(self):
+        '''
+        Test whether Scheduler.set_sched_config() works as expected.
+        '''
+        self.our_sched = self.scheds['default']
+
+        # First, get default settings for a few values and verify
+        # they are what they should be as of this version of test.
+        slines = self.set_and_get({})
+        def_rr = self.get_a_setting(slines, 'round_robin')
+        def_nsk = self.get_a_setting(slines, 'node_sort_key')
+        def_jsk = self.get_a_setting(slines, 'job_sort_key')
```

Comment:
Why aren't you using the sched_config dictionary? Then these can just be:
def_rr = self.scheduler.sched_config['round_robin']
def_nsk = self.scheduler.sched_config['node_sort_key']
def_jsk = self.scheduler.sched_config['job_sort_key']

Reply:
That section of code serves two functions. First, the obvious one of fetching the default values. The second function was to validate that the code used in set_and_get() and get_a_setting() returned values the way I expected (e.g., as lists of strings).

I changed the test to use the sched_config dictionary when checking the default values, which removes a dependency the original code had on knowing those defaults.

You know, I've spent a lot more time writing the PTL tests than I did chasing down and fixing the bug. :-)

Reply:
> You know, I've spent a lot more time writing the PTL tests than I did chasing down and fixing the bug. :-)

That would be a failure in PTL's part, but I still think that you can utilize PTL's built-in functionality to write the test as:

```
class TestSchedConfig(TestSelf):
    def test_set_sched_config(self):
        our_sched = self.scheds['default']
        sched_conf = our_sched.sched_config
        nsk = sched_conf['node_sort_key']
        jsk = sched_conf['job_sort_key']

        # See if we can change an existing value, and leave others alone
        our_sched.set_sched_config({'round_robin': 'True ALL'}, apply=True)
        our_sched.parse_sched_config()
        rr = sched_conf['round_robin']
        self.assertEqual(rr, ["True ALL"])
        self.assertEqual(sched_conf['node_sort_key'], nsk, "Change to unexpected setting")
        self.assertEqual(sched_conf['job_sort_key'], jsk, "Change to unexpected setting")

        # Repeat for initially missing setting
        our_sched.set_sched_config({'job_sort_key': '"ncpus HIGH"'}, apply=True)
        our_sched.parse_sched_config()
        jsk = sched_conf['job_sort_key']
        self.assertEqual(jsk, ['"ncpus HIGH"'])
        self.assertEqual(sched_conf['round_robin'], rr, "Change to unexpected setting")
        self.assertEqual(sched_conf['node_sort_key'], jsk, "Change to unexpected setting")

.............
```

So, I don't think you need the helper functions (set_and_get() and get_a_setting()) at all, let PTL do the parsing and caching for you.